align_system/utils/distribution_matching_utils.py

```diff
@@ -1,22 +1,46 @@
 from align_system.utils import kde_utils
+import numpy as np
 
-
-def match_to_target_kde_sample(alignment_target, available_actions, kde_norm):
+def match_to_target_kde_sample(alignment_target, available_actions, kde_norm, num_kde_samples=1):
 	'''
-	Samples a random value from the target KDE and selects the action with KDMA value closest to the sample
+	Returns the choice with min average distance to random samples from the target KDE
 	'''
 	target_kdma = alignment_target.kdma_values[0] # TODO extend this to multi-KDMAs
 	target_kde = kde_utils.load_kde(target_kdma, kde_norm)
-	sampled_target = target_kde.sample(1)[0]
+	sampled_targets = target_kde.sample(num_kde_samples)
 	
 	selected_action = None
 	min_dist = float('inf')
 	for action in available_actions:
 		if action.kdma_association is not None:
+			sum_of_dists = 0
 			choice_value = action.kdma_association[target_kdma['kdma']]
-			dist = abs(sampled_target-choice_value)
+			for sampled_target in sampled_targets:
+				sum_of_dists += abs(sampled_target-choice_value)
+			dist = sum_of_dists/num_kde_samples
 			if dist < min_dist:
 				min_dist = dist 
 				selected_action = action
 
+	return selected_action
+
+
+def max_likelihood_matching(alignment_target, available_actions, kde_norm):
+	'''
+	MLE - Returns the choice with the max likelihood under the target KDE
+	'''
+	selected_action = None
+	target_kdma = alignment_target.kdma_values[0] # TODO extend this to multi-KDMAs
```

Comment:
I know we keep leaving these as TODOs, but I'm wondering where we're at the point where we're starting to get so many that we should actually check to confirm that we haven't been provided a multi-kdma target so that we aren't giving ourselves a footgun

Reply:
Good point, I added a runtime error

Reply:
Do we need to do the same thing for any other ADMs?

Reply:
Just added for the comparative regression ADM